packages/rOpenScPCA/R/sweep-clusters.R

```diff
@@ -0,0 +1,126 @@
+#'   to vary this parameter.
+#' @param seed Random seed to set for clustering.
+#' @param threads Number of threads to use. Default is 1.
+#' @param pc_name Name of principal components slot in provided object. This argument is only used
+#'   if a SingleCellExperiment or Seurat object is provided. If not provided, the SingleCellExperiment
+#'   object name will default to "PCA" and the Seurat object name will default to "pca".
+#'
+#' @return A list of data frames from performing clustering across all parameter combinations.
+#'   Columns include `cluster_set` (identifier column for results from a single clustering run),
+#'   `cell_id`, and `cluster`. Additional columns represent algorithm parameters and include at least:
+#'   `algorithm`, `weighting`, and `nn`. Louvain and leiden clustering will also include `resolution`,
+#'   and leiden clustering will further include `objective_function`.
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # perform louvain clustering with jaccard weighting (defaults),
+#' # varying the nearest neighobor parameter.
+#' cluster_df <- sweep_clusters(sce_object, nn = c(10, 15, 20, 25))
+#'
+#' # perform louvain clustering, with jaccard and rank weighting, and
+#' # varying the nearest neighbor and resolution parameters.
+#' cluster_df <- sweep_clusters(
+#'   sce_object,
+#'   algorithm = "louvain",
+#'   weighting = c("jaccard", "rank"),
+#'   nn = c(10, 15, 20, 25),
+#'   resolution = c(0.5, 1)
+#' )
+#'
+#' # perform walktrap and louvain clustering with jaccard weighting, and
+#' # varying the nearest neighbors for both algorithms, and resolution for louvain.
+#' cluster_df <- sweep_clusters(
+#'   sce_object,
+#'   algorithm = c("walktrap", "louvain"),
+#'   weighting = "jaccard",
+#'   nn = c(10, 15, 20, 25),
+#'   resolution = c(0.5, 1)
+#' )
+#' }
+sweep_clusters <- function(
+    x,
+    algorithm = "louvain",
+    weighting = "jaccard",
+    nn = 10,
+    resolution = 1, # louvain or leiden
+    objective_function = "CPM", # leiden only
+    threads = 1,
+    seed = NULL,
+    pc_name = NULL) {
+  # Ensure input is a matrix for slightly faster processing later
+  if (any(class(x) %in% c("matrix", "Matrix"))) {
+    stopifnot(
+      "The matrix must have row names representing cell ids, e.g. barcodes." = is.character(rownames(x))
+    )
+  } else if (is(x, "SingleCellExperiment") || is(x, "Seurat")) {
+    x <- extract_pc_matrix(x, pc_name = pc_name)
+  } else {
+    stop("The first argument should be one of: a SingleCellExperiment object, a Seurat object, or a matrix with row names.")
+  }
+
+  # Collect all specific inputs into a single list
+  sweep_params <- tidyr::expand_grid(
+    algorithm = unique(algorithm),
+    weighting = unique(weighting),
+    nn = unique(nn),
+    resolution = unique(resolution),
+    objective_function = unique(objective_function)
+  ) |>
+    # set unused parameters for the given algorithm to their defaults
```

Comment:
I just want to clarify the reasoning here
```suggestion
    # set unused parameters for each algorithm to default; this will allow duplicates to be removed by distinct()
```